lib/src/main/kotlin/com/swmansion/starknet/provider/rpc/JsonRpcResponse.kt

```diff
@@ -35,23 +30,29 @@ internal data class JsonRpcError(
     val message: String,
 )
 
-internal class JsonRpcResponseDeserializer<T>(private val dataSerializer: KSerializer<T>) : DeserializationStrategy<T> {
-    override val descriptor: SerialDescriptor
-        get() = dataSerializer.descriptor
+internal fun <T> buildJsonHttpDeserializer(deserializationStrategy: KSerializer<T>): HttpRequestDeserializer<T> {
+    return { response ->
+        if (!response.isSuccessful) {
+            throw if (response.body == null) RequestFailedException("Request failed") else RequestFailedException(
+                response.body,
+            )
+        }
 
-    override fun deserialize(decoder: Decoder): T {
-        require(decoder is JsonDecoder)
+        val jsonRpcResponse =
+            Json.decodeFromString(
+                JsonRpcResponse.serializer(deserializationStrategy),
+                response.body!!,
+            ) // Can we assume that if the request was successful, it will have a body?
 
-        val responseJson = decoder.decodeJsonElement()
-        val format = Json { ignoreUnknownKeys = true }
-        val response = format.decodeFromJsonElement(JsonRpcResponse.serializer(dataSerializer), responseJson)
+        if (jsonRpcResponse.error != null) {
+            throw RpcRequestFailedException(jsonRpcResponse.error.code, jsonRpcResponse.error.message)
+        }
 
-        if (response.result != null) {
-            return response.result
-        } else if (response.error != null) {
-            throw RpcRequestFailedException(response.error.code, response.error.message)
-        } else {
-            throw RequestFailedException()
+        if (jsonRpcResponse.result == null) {
+            // FIXME add more specific error
+            throw RequestFailedException(response.body)
```

Comment:
Maybe we could also add a string explaining what's wrong? LIke "Empty result" in this case?